File: bfs/the_maze.rb
# Approach 1: BFS
def has_path(maze, src, dst)
    @maze = maze
    q, @m, @n = [start], @maze.size, @maze[0].size
    dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]] # [right, left, down, up]

    while !q.empty?
        i, j = q.shift()
        maze[i][j] = 2
        return true if [i, j] == destination

        dirs.each do |x, y|
            r, c = i, j
            while valid?(r + x, c + y)
                r += x; c += y
            end
            q.push([r, c]) if @maze[r][c].zero?
        end
    end

    false
end

def valid?(r, c)
    r.between?(0, @m - 1) && c.between?(0, @n - 1) && @maze[r][c] != 1
end

require 'set'
# Approach 2: DFS
def has_path_dfs(maze, src, dst)
    @maze, @m, @n = maze, maze.size, maze[0].size 
    dfs(src, dst)
end

DIRS = [[0, 1], [0, -1], [1, 0], [-1, 0]] # [right, left, down, up]
def dfs(s, d)
    i, j = s
    return false if @maze[i][j] == 2
    @maze[i][j] = 2
    return true if s == d

    DIRS.each do |x, y|
        r, c = s
        while valid?(r + x, c + y)
            r += x; c += y
        end
        return true if dfs([r, c], d)
    end

    false
end

def valid?(r, c)
    r.between?(0, @m - 1) && c.between?(0, @n - 1) && @maze[r][c] != 1
end

# 490. The Maze
# https://leetcode.com/problems/the-maze/description/

# Apporach 1: BFS, Time: O(m * n), Space: O(m * n)
# Approach 2: DFS, Time: O(m * n), Space: O(m * n)


require 'test/unit'
extend Test::Unit::Assertions

# BFS Tests
maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path(maze, [0,4], [3,2]), false)
maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path(maze, [0,4], [4,4]), true)

# DFS Tests
maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path_dfs(maze, [0,4], [3,2]), false)
maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path_dfs(maze, [0,4], [4,4]), true)

File: bfs/the_maze.py
from typing import List


# Approach 1: DFS


# Approach 2: BFS
class Solution:
    def hasPath(self, maze: List[List[int]], src: List[int], dst: List[int]) -> bool:
        if src == dst: return True
        m, n = len(maze), len(maze[0])
        q, maze[src[0]][src[1]] = [src], 2
        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        while q:
            i, j = q.pop(0)

            for dx, dy in dirs:
                r, c = i, j
                while self.not_wall(maze, m, n, r + dx, c + dy): r += dx; c += dy

                if maze[r][c] == 2: continue
                if [r, c] == dst: return True

                q.append([r, c])
                maze[r][c] = 2

        return False

    def not_wall(self, maze, m, n, i, j):
        return 0 <= i <= m - 1 and 0 <= j <= n - 1 and maze[i][j] != 1


# 490. The Maze
# https://leetcode.com/problems/the-maze/description/
